### PR TITLE
Add a confirm attribute for recipes

### PR DESCRIPTION
**Pull request by Hwatwasthat**
This would add a "Confirm" attribute that requests confirmation before running a recipe. Runs prior to evaluation, and defaults to not running. Potential solution to issue [#1711](https://github.com/casey/just/issues/1711).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,4 @@
 /test-utilities/Cargo.lock
 /test-utilities/target
 /tmp
+.vscode/
```
Comment (casey) [inline .gitignore]: This should be:

```suggestion
/.vscode
```

Also, the lines in this file should be sorted.

File: src/attribute.rs [hunk 2]
```diff
@@ -13,6 +13,7 @@ pub(crate) enum Attribute {
   Private,
   Unix,
   Windows,
+  Confirm
```
Comment (casey) [inline src/attribute.rs]: These enum variants should be sorted.

File: src/recipe.rs [hunk 3]
```diff
@@ -63,6 +63,16 @@ impl<'src, D> Recipe<'src, D> {
     self.name.line
   }
 
+  pub(crate) fn confirm(&self) -> bool {
```
Comment (casey) [inline src/recipe.rs]: This function should be fallible, since `read_line` can fail, so it should return `RunResult<bool>`

File: src/recipe.rs [hunk 4]
```diff
@@ -159,6 +175,8 @@ impl<'src, D> Recipe<'src, D> {
         }
       }
 
+      
+
       if comment_line {
```
Comment (casey) [inline src/recipe.rs]: Extra whitespace:

```suggestion
      if comment_line {
```

File: src/recipe.rs [hunk 5]
```diff
@@ -63,6 +63,16 @@ impl<'src, D> Recipe<'src, D> {
     self.name.line
   }
 
+  pub(crate) fn confirm(&self) -> bool {
+    if self.attributes.contains(&Attribute::Confirm) {
+      eprint!("Confirm running {} (y/N): ", self.name);
+      let mut buf = String::new();
+      std::io::stdin().read_line(&mut buf).is_ok() && buf.trim_end().eq_ignore_ascii_case("y")
```
Comment (casey) [inline src/recipe.rs]:
```suggestion
      let mut line = String::new();
      std::io::stdin().read_line(&mut line)?;
      let line = line.trim().to_lowercase();
      line == "y" || line == "yes"
```

**Review by casey — CHANGES_REQUESTED** [review]
I just added some comments, and I thought of something else: If a recipe doesn't run, any recipe in the current invocation which depends on that recipe should also not run.

**Comment by Hwatwasthat** [comment]
> I just added some comments, and I thought of something else: If a recipe doesn't run, any recipe in the current invocation which depends on that recipe should also not run.

I've added in the confirmation during the running of dependencies. Now, if a recipe that requires confirmation is run as a dependency and is not confirmed then it will return an error of type `DependencyNotConfirmed`, which I could only find a way to insert the name of the dependency being run, I can't find an easy way to inject the parents name without adding in a lot more convolution.

The confirmation is carried out at the start of the `Justfile::run_recipe` function, as I couldn't see another place to put it that would prevent anything running and would allow the returning of an error rather than the ok of a single recipe that is being run (as that would stop any later recipes that were from running as I understand it, which is not the intended use of confirmation (again, as I understand it).

**Comment by Hwatwasthat** [comment]
Apologies, on testing a bit I realised that I've added in a double confirm. I'll figure out the best way to resolve this and then update the pull.

**Comment by casey** [comment]
Aside from the double confirm, that all sounds good. I think returning an error if a recipe isn't confirmed is sensible. I think that if we get to a recipe that needs confirmation, and it isn't confirmed, we can abort the whole run, and not run any other recipes. Although it is true that we could run any recipes which aren't dependencies of the unconfirmed recipe, not running them doesn't seem like a big deal.

Also, this will require a couple of tests: One that a recipe that requires and receives confirmation runs, and one that a recipe which requires confirmation but doesn't receive it does not run.

**Comment by casey** [comment]
Also, if you feel like adding a `--yes` flag that confirms all recipes (per this comment https://github.com/casey/just/issues/1711#issuecomment-1807732753) which seems useful, go for it, otherwise it can be left to another PR.

**Comment by Hwatwasthat** [comment]
> Aside from the double confirm, that all sounds good. I think returning an error if a recipe isn't confirmed is sensible. I think that if we get to a recipe that needs confirmation, and it isn't confirmed, we can abort the whole run, and not run any other recipes. Although it is true that we could run any recipes which aren't dependencies of the unconfirmed recipe, not running them doesn't seem like a big deal.
> 
> Also, this will require a couple of tests: One that a recipe that requires and receives confirmation runs, and one that a recipe which requires confirmation but doesn't receive it does not run.

Ok, that makes things easier then, I've got a version working that does that so I'll just plow on with it. Tests to come with that in the next commit.

**Comment by Hwatwasthat** [comment]
> Also, if you feel like adding a `--yes` flag that confirms all recipes (per this comment [#1711 (comment)](https://github.com/casey/just/issues/1711#issuecomment-1807732753)) which seems useful, go for it, otherwise it can be left to another PR.

I was just wondering on the name. yes is a bit ambiguous, maybe `--force-confirm` (just to make it clear), I feel like `--force` or `--no-confirm` doesn't work, as force is ambiguous and no confirm reads to me as the inverse (automatically denying confirm stuff). I have a flag working, right now it's `--yes`.

**Review by casey — APPROVED** [review]
I tweaked a few error messages, as well as printed the error confirmation message to stderr, which is also line-buffered, so no need to flush.

Thanks for the PR!

**Pull request merged**